### PR TITLE
Add Gaussian option to SmoothMD

### LINKED ISSUE #14731: Implement a Gaussian option for SmoothMD
Currently only a "hat" option is implemented. Russell Ewings has mentioned that Gaussian is also important.

"Regarding the Gaussian smoothing (which should only apply to MDHisto workspaces, by the way) we define the width with an array that has the same number of elements as there are dimensions in the workspace. The nth element of this array corresponds to the full-width-half-maximum of the Gaussian along the nth dimension of the data. Crucially, this width is given in terms of number of bins, rather than Q or energy units. The truncation of the Gaussian is done when the value in the wings falls below 0.02 of the peak."

### PR DESCRIPTION
**Pull request by matthew-d-jones**
Closes #14731 

Added a "Gaussian" function option for the SmoothMD algorithm and updated the documentation. The "WidthVector" is now a list of `double` rather than `int`, but validators check that only odd integer values are given for the existing "Hat" function case. Widths for the Gaussian case are given as the FWHM in units of pixels, which is consistent with Horace.

As the multidimensional Gaussian function is linearly separable I carry out the convolution by using a 1D kernel in each dimension. This reduces the overall number of calculations necessary.

The Gaussian filter is created by integrating the function over the width of each pixel, to give a more accurate result than just evaluating it at the centre position of the pixel, this was cheap using `std::erf`. The filter is truncated at the point where the pixel value falls below 0.02, which is consistent with Horace.

**To Test:**

Here is a qualitative test, which should work with any png file (I tried it with https://en.wikipedia.org/wiki/Mantidae#/media/File:Zorak-Mantis.png):

``` python
# Image processing in Mantid! (Gaussian blur)
import numpy as np
import matplotlib.image as mpimg
img = mpimg.imread('/path/to/image.png')
img = np.mean(img, 2)
ws_img_in = CreateMDWorkspace(Dimensions='2', Extents='-2,2,-2,2', Names='h,k',
                       Units='rlu,rlu', SplitInto='13')
ws_img_in = BinMD(ws_img_in,AlignedDim0='h,-2,2,'+str(img.shape[0]),AlignedDim1='k,-2,2,'+str(img.shape[1]))
ws_img_in.setSignalArray(img.astype(float))
ws_smooth_img = SmoothMD(InputWorkspace=ws_img_in, Function='Gaussian', WidthVector='11')
# Compare ws_img_in and ws_smooth_img in the slice viewer (use Norm="no" and set colourbar range to 0 to 1)
```

### CODE REVIEW

**Comment by AndreiSavici** [comment]
A couple of caveats:
1. The Gaussian widths that you provide are along the dimensions of the workspace, so one has to be careful when looking at non-orthogonal lattices. 
2. The multi-dimensional Gaussian is in general separable into 1D Gaussians, but they might not be (in general) aligned with the dimensions. A 2D Gaussian, in the most general form is A*exp(a*x^2+b*x*y+c*y^2)
3. Is the Horace way (pixels) the best way to introduce the smoothing? I would guess that one might want to give sigmas in rlu or inverse angstroms.

**Comment by matthew-d-jones** [comment]
@AndreiSavici @quantumsteve 
You raise a good point, this only deals with the ws dimension/lattice aligned case. I implemented this in response to requests to have an equivalent to Horace's Gaussian smoothing.

What I've added to the SmoothMD algorithm does not impede implementing the more general case in the future. I also think that if the general case was implemented we are likely to still want to use the 1D method, when possible, to reduce computation time. So I think we can treat this as v1.

**Comment by quantumsteve** [comment]
I think that a more general Gaussian function would be very useful in Mantid. If you do assume separable 1D Gaussians and orthogonal axes, please make sure it fails when the input doesn't match these assumptions.

File: Framework/DataObjects/src/MDHistoWorkspaceIterator.cpp [hunk 1]
```diff
@@ -679,6 +679,59 @@ std::vector<size_t> MDHistoWorkspaceIterator::findNeighbourIndexesByWidth(
...
+    neighbourIndexes[i] = neighbour_index;
+    if (neighbour_index < m_ws->getNPoints() &&
+        Utils::isNeighbourOfSubject(m_nd, neighbour_index, m_index,
+                                    m_indexMaker, m_indexMax, widths)) {
+      indexValidity[i] = true;
+    }
+  }
+
+  return std::make_pair(neighbourIndexes, indexValidity);
+}
```
Comment (AntonPiccardoSelg) [inline Framework/DataObjects/src/MDHistoWorkspaceIterator.cpp]: a couple of instances where `auto` could have been used

File: Framework/MDAlgorithms/src/SmoothMD.cpp [hunk 2]
```diff
@@ -48,21 +53,103 @@ namespace {
...
+  const double sigma = (fwhm * 0.42463) / 2.0;
+  const double sigma_factor = M_SQRT1_2 / (fwhm * 0.42463);
+
+  KernelVector kernel_one_side;
+  // Start from centre and calculate values going outwards until value < 0.02
+  // Use erf to get the value of the Gaussian integrated over the width of the
+  // pixel, more accurate than just using centre value of pixel and erf is fast
+  double pixel_value = std::erf(0.5 * sigma_factor) * sigma;
+  int pixel_count = 0;
+  while (pixel_value > 0.02) {
```
Comment (AntonPiccardoSelg) [inline Framework/MDAlgorithms/src/SmoothMD.cpp]: Is 0.02 an arbitrary cutoff? or is there a reason behind using this value?

File: Framework/MDAlgorithms/src/SmoothMD.cpp [hunk 3]
```diff
@@ -186,6 +281,126 @@ SmoothMD::hatSmooth(IMDHistoWorkspace_const_sptr toSmooth,
...
+ * @param weightingWS : Weighting workspace (optional)
+ * @return Smoothed MDHistoWorkspace
+ */
+IMDHistoWorkspace_sptr
+SmoothMD::gaussianSmooth(IMDHistoWorkspace_const_sptr toSmooth,
+                         const WidthVector &widthVector,
+                         OptionalIMDHistoWorkspace_const_sptr weightingWS) {
+
+  const bool useWeights = weightingWS.is_initialized();
+  uint64_t nPoints = toSmooth->getNPoints();
```
Comment (AntonPiccardoSelg) [inline Framework/MDAlgorithms/src/SmoothMD.cpp]: again some cases where `auto` could have been used, also below

File: Framework/MDAlgorithms/src/SmoothMD.cpp [hunk 4]
```diff
@@ -186,6 +281,126 @@ SmoothMD::hatSmooth(IMDHistoWorkspace_const_sptr toSmooth,
...
+          write_ws->setSignalAt(iteratorIndex, sumSignal);
+          write_ws->setErrorSquaredAt(iteratorIndex, sumSquareError);
+        }
+        progress.report();
+
+      } while (iterator->next());
+      PARALLEL_END_INTERUPT_REGION
+    }
+    PARALLEL_CHECK_INTERUPT_REGION
+  }
```
Comment (AntonPiccardoSelg) [inline Framework/MDAlgorithms/src/SmoothMD.cpp]: seems like a very long method. should we have anonymous functions which can do parts of the work and make it more readable?

**Comment by AntonPiccardoSelg** [comment]
* Code review: code looks fine. There is one method which should probably be split further up. We agreed that this should be part of #15847 . The code is well tested.
* documentation: the user side documentation is fine, release notes have been added.
* functional testing: played around with some suggested images. Behaves as expected.

Once the builds have passed, I am happy to :shipit:

**Comment by FedeMPouzols** [comment]
Looks pretty solid and more improvements follow.
Also great that the documentation gives specific equations. More equations and references were one of the requests at the last annual meeting.

**Pull request merged**